test/rulesets/LinearAlgebra/symmetric.jl

```diff
@@ -283,4 +283,225 @@
+                    @test typeof(Y_ad) === typeof(Y)
+                    :uplo in propertynames(Y) && @test Y_ad.uplo == Y.uplo
+                    @test ∂Y_ad isa typeof(Y)
+                    :uplo in propertynames(∂Y_ad) && @test ∂Y_ad.uplo == Y.uplo
+                    @test parent(∂Y_ad) ≈ jvp(_fdm, x -> parent(f(TA(x, uplo))), (A.data, ΔA.data))
+
+                    @test frule((Zero(), Zero()), f, A) == (Y, Zero())
+                end
+
+                @testset "stable for (almost-)singular input" begin
+                    λ, U = eigen(rand_matfun_input(f, TA, T, :U, n, true))
+                    m = div(n, 2)
+                    λ[1:m] .= λ[m+1:2m] .+ cbrt(eps(eltype(λ))) / 100
+                    A = TA(U * Diagonal(λ) * U')
+                    ΔA = TA(randn(T, n, n))
+                    _, ∂Y = frule((Zero(), ΔA), f, A)
+                    @test parent(∂Y) ≈ jvp(_fdm, x -> parent(f(TA(x))), (A.data, ΔA.data))
+
+                    λ[1:m] .= λ[m+1:2m]
+                    A2 = TA(U * Diagonal(λ) * U')
+                    ΔA2 = TA(randn(T, n, n))
+                    _, ∂Y2 = frule((Zero(), ΔA2), f, A2)
+                    @test parent(∂Y2) ≈ jvp(_fdm, x -> parent(f(TA(x))), (A2.data, ΔA2.data))
+                end
+
+                f ∉ (log,sqrt,acosh) && @testset "low-rank matrix" begin
+                    λ, U = eigen(rand_matfun_input(f, TA, T, :U, n, true))
+                    λ[2:n] .= 0
+                    A = TA(U * Diagonal(λ) * U')
+                    ΔA = TA(randn(T, n, n))
+                    _, ∂Y = frule((Zero(), ΔA), f, A)
+                    @test parent(∂Y) ≈ jvp(_fdm, x -> parent(f(TA(x))), (A.data, ΔA.data))
+                end
+            end
+
+            @testset "rrule" begin
+                @testset for uplo in (:L, :U), hermout in (true, false)
+                    A = rand_matfun_input(f, TA, T, uplo, n, hermout)
+                    Y = f(A)
+                    istypestable = try
+                        @inferred f(A)
+                        true
+                    catch ErrorException
+                        false
+                    end
+                    ΔY = if Y isa Matrix
+                        randn(eltype(Y), n, n)
+                    else
+                        typeof(Y)(randn(eltype(Y), n, n), Y.uplo)
+                    end
+                    if istypestable
+                        Y_ad, back = @inferred rrule(f, A)
+                    else
+                        TY = if T <: Real
+                            Union{Symmetric{Complex{T}},Symmetric{T}}
+                        else
+                            Union{Matrix{T},Hermitian{T}}
+                        end
+                        Y_ad, back = @inferred Tuple{TY,Any} rrule(f, A)
+                    end
+                    @test Y_ad == Y
+                    @test typeof(Y_ad) === typeof(Y)
+                    :uplo in propertynames(Y) && @test Y_ad.uplo == Y.uplo
```

Comment:
Sigh `hasproperty` is not in 1.0.
Is it in Compat.jl?

Reply:
It does, although for some reason using `hasproperty` still caused an error. I'll try again.